### PR TITLE
Openshift Registry : Add project/group/user objects

### PR DESCRIPTION
**Pull request by sub-mod**
https://github.com/cockpit-project/cockpit/pull/3499

### CODE REVIEW

**Comment by stefwalter** [comment]
I would suggest rebasing on top of #3506

File: pkg/registry/scripts/projects.js [hunk 1]
```diff
@@ -98,7 +105,39 @@
...
+                });
+            }
+
+            return {
+                create: createDialog,
+            };
+        }
+    ])
+
+    .factory('userActions', [
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: Lets just put this into projectActions. That is: the actions for the projects page. And not have separate userActions, groupActions.

File: pkg/registry/scripts/projects.js [hunk 2]
```diff
@@ -98,7 +105,39 @@
             function createDialog() {
                 return $modal.open({
                     controller: 'ProjectNewCtrl',
-                    templateUrl: 'views/project-new-dialog.html',
+                    templateUrl: 'views/project/add-project-dialog.html',
+                });
+            }
+
+            return {
+                create: createDialog,
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: We can rename this to createProject.

File: pkg/registry/scripts/projects.js [hunk 3]
```diff
@@ -128,7 +167,13 @@
                 var display = fields.display.trim();
                 var description = fields.description.trim();
 
-                if (!NAME_RE.test(name)) {
+                if (!name) {
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: Doesn't the NAME_RE already handle the case of empty names? If not, lets just adjust NAME_RE to not match empty strings.

File: pkg/registry/scripts/projects.js [hunk 4]
```diff
@@ -152,5 +197,96 @@
...
+    ])
+
+    .controller('UserNewCtrl', [
+        '$q',
+        '$scope',
+        "kubeMethods",
+        function($q, $scope, methods) {
+            var fields = {
+                name: "",
+                password: ""
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: I think this dialog isn't supposed to have a password field. Should it be an identity field instead?

File: pkg/registry/scripts/projects.js [hunk 5]
```diff
@@ -152,5 +197,96 @@
...
+            };
+
+            $scope.fields = fields;
+
+            $scope.performCreate = function performCreate() {
+                var defer, ex;
+
+                var name = fields.name.trim();
+                var password = ["anypassword:" + fields.password.trim()];
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: See above. I think we should be typing in a full identity string here.

File: pkg/registry/scripts/projects.js [hunk 6]
```diff
@@ -152,5 +197,96 @@
...
+
+                if (!name) {
+                    ex = new Error("Name cannot be empty.");
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: Same as above, lets make NAME_RE handle this case.

File: pkg/registry/scripts/projects.js [hunk 7]
```diff
@@ -152,5 +197,96 @@
...
+
+            $scope.fields = fields;
+
+            $scope.performCreate = function performCreate() {
+                var defer, ex;
+
+                var name = fields.name.trim();
+
+                if (!name) {
+                    ex = new Error("Name cannot be empty.");
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: Same as above with regards to NAME_RE.

File: pkg/registry/styles/projects.less [hunk 8]
```diff
@@ -1,6 +1,9 @@
 
 @tag-border: @link-color;
 
+#add-entity {
```
Comment (stefwalter) [inline pkg/registry/styles/projects.less]: Why is this needed?

File: pkg/registry/views/projects-page.html [hunk 9]
```diff
@@ -52,8 +56,12 @@ <h3 translatable="yes">Groups</h3>
 
         <thead ng-if="!quiet">
             <tr>
-                <td colspan="4">
-                    <h3 translatable="yes">Users</h3>
+                <td colspan="3">
+                    <a class="pull-right" id="add-entity" ng-click="addUserDialog()">
+                        <i translatable="yes" class="pficon pficon-add-circle-o"></i>
+                        <span translatable="yes">New User</span>
```
Comment (stefwalter) [inline pkg/registry/views/projects-page.html]: This should say 'Add user'

File: pkg/registry/views/project/add-user-dialog.html [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+                </td>
+            </tr>
+            <tr>
+                <td class="top">
+                    <label class="control-label" for="password" translatable="yes">
+                        Identity
+                    </label>
+                </td>
+                <td>
+                    <input id="password" ng-model="fields.password"class="form-control" type="text">
```
Comment (stefwalter) [inline pkg/registry/views/project/add-user-dialog.html]: Missing space here.

File: pkg/registry/views/project/add-user-dialog.html [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+                </td>
+            </tr>
+        </table>
+    </div>
+    <div class="modal-footer">
+        <button class="btn btn-default btn-cancel">Cancel</button>
+        <button class="btn btn-primary" ng-click="complete(performCreate())">Add</button>
```
Comment (stefwalter) [inline pkg/registry/views/project/add-user-dialog.html]: Needs to be translatable.

File: pkg/registry/views/project/add-user-dialog.html [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+                    </label>
+                </td>
+                <td>
+                    <input id="password" ng-model="fields.password"class="form-control" type="text">
+                </td>
+            </tr>
+        </table>
+    </div>
+    <div class="modal-footer">
+        <button class="btn btn-default btn-cancel">Cancel</button>
```
Comment (stefwalter) [inline pkg/registry/views/project/add-user-dialog.html]: Needs to be translatable.

File: pkg/registry/views/project/add-user-dialog.html (hunk 10 above)
Comment (stefwalter) [inline pkg/registry/views/project/add-user-dialog.html]: Lets call this field 'identity'

File: pkg/registry/Makefile.am [hunk 13]
```diff
@@ -11,7 +11,9 @@ registry_VIEWS = \
 	pkg/registry/views/imagestream-meta.html \
 	pkg/registry/views/imagestream-page.html \
 	pkg/registry/views/imagestream-panel.html \
-	pkg/registry/views/project-new-dialog.html \
+	pkg/registry/views/project/add-group-dialog.html \
```
Comment (stefwalter) [inline pkg/registry/Makefile.am]: I think we should keep a flat directory hierarchy here, and name these files like this:

project-new-dialog.html
group-new-dialog.html
user-add-dialog.html

File: pkg/registry/scripts/projects.js [hunk 14]
```diff
@@ -70,7 +70,9 @@
         '$modal',
         'projectUtil',
         'projectActions',
-        function($scope, loader, select, $modal, util, actions) {
+        'userActions',
+        'groupActions',
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: As noted below, these should all be part of projectActions

**Comment by stefwalter** [comment]
Thanks for posting this pull request. Mosttly good, just a few comments.

File: pkg/registry/scripts/projects.js (hunk 3 above)
Comment (sub-mod) [inline pkg/registry/scripts/projects.js]: wanted to show a "Name cannot be empty" error message
to show that it is a mandatory field

File: pkg/registry/styles/projects.less (hunk 8 above)
Comment (sub-mod) [inline pkg/registry/styles/projects.less]: other add buttons  still dont align, this only aligns Project(the first button)

File: pkg/registry/Makefile.am (hunk 13 above)
Comment (sub-mod) [inline pkg/registry/Makefile.am]: why flat  ?

File: pkg/registry/scripts/projects.js (hunk 3 above)
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: Alright. Makes sense. But in that case please refactor this code so that it's not a big set of copy paste. We should be creating Errors based on conditions, and if an error is set, have one block of code to create the deferred.

In addition this, most of code should be factored out into a function, called validate-name that returns an exception ... so that it's not duplicated between all these dialogs. Because you're now combining all of these into one projectActions block, that should be easy.

File: pkg/registry/styles/projects.less (hunk 8 above)
Comment (stefwalter) [inline pkg/registry/styles/projects.less]: It's illegal HTML to have duplicate ids on the same page. But I'll take a look at how to solve this better.

File: pkg/registry/Makefile.am (hunk 13 above)
Comment (stefwalter) [inline pkg/registry/Makefile.am]: Because we already have nested directories for the different types of files. Much like the openshift project. If we had chosen to put all the project stuff in one directory (scripts, views, styles) then this would make sense, but we've chosen to have directories by asset type.

**Comment by stefwalter** [comment]
```
pkg/registry/scripts/projects.js:163:21: 'ex' is already defined.
pkg/registry/scripts/projects.js:207:21: 'ex' is already defined.
pkg/registry/scripts/projects.js:243:21: 'ex' is already defined.
```
